geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,391 @@
+    final Set<?> entries;
+    if (region instanceof PartitionedRegion && sender.isParallel()) {
+      entries =
+          ((PartitionedRegion) region).getDataStore().getAllLocalBucketRegions()
+              .stream()
+              .flatMap(br -> ((Set<?>) br.entrySet()).stream()).collect(Collectors.toSet());
+    } else {
+      entries = region.entrySet();
+    }
+
+    final InternalCache cache = (InternalCache) context.getCache();
+    int replicatedEntries = 0;
+    final long startTime = clock.millis();
+
+    Iterator<?> iter = entries.iterator();
+
+    while (iter.hasNext()) {
+      int batchIndex = 0;
+      List<GatewaySenderEventImpl> batch = new ArrayList<>();
+      while (iter.hasNext() && batchIndex < batchSize) {
+        GatewaySenderEventImpl event = createGatewaySenderEvent(cache, (InternalRegion) region,
+            sender, (Region.Entry) iter.next());
+        if (event != null) {
+          batch.add(event);
+          batchIndex++;
+        }
+      }
+      try {
+        sendBatch(sender,
+            replicatedEntries, maxRate, startTime, batch);
+        replicatedEntries += batch.size();
+      } catch (InterruptedException e) {
+        return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+            "Operation canceled after having replicated " + replicatedEntries + " entries");
+      } catch (BatchException70 e) {
+        return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+            "Error (" + e.getMessage() + ") in operation after having replicated "
+                + replicatedEntries + " entries");
+      }
+    }
+
+    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+        "Entries replicated: " + replicatedEntries);
+  }
+
+  private GatewaySenderEventImpl createGatewaySenderEvent(InternalCache cache,
+      InternalRegion region, GatewaySender sender, Region.Entry entry) {
+    final EntryEventImpl event;
+    if (region instanceof PartitionedRegion) {
+      event = createEventForPartitionedRegion(sender, cache, region, entry);
+    } else {
+      event = createEventForReplicatedRegion(cache, region, entry);
+    }
+    if (event == null) {
+      return null;
+    }
+    try {
+      return new GatewaySenderEventImpl(EnumListenerEvent.AFTER_CREATE, event, null, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  private void sendBatch(GatewaySender sender,
+      int replicatedEntries, long maxRate, long startTime,
+      List batch) throws InterruptedException, BatchException70 {
+    int batchId = 1971;
+    PoolImpl senderPool = ((AbstractGatewaySender) sender).getProxy();
+    if (senderPool == null) {
+      throw new BatchException70("No connection pool available towards receiver", null, 0, batchId);
+    }
+    Connection connection = senderPool.acquireConnection();
```

Comment:
This call is acquiring a Connection for every batch, but it isn't returning it to the pool, so if you have 100 batches to be sent, this will create 100 Connections to the remote site.

I refactored it a bit to acquire a Connection before sending any batches, then returning it after all the batches are sent. I also set it to be destroyed upon return. I also set the batchId to 0 and incremented it each time through the loop. That eliminates the warnings on the remote site.

Here is my modified code. Feel free to use it however you want.
```
private CliFunctionResult replicateRegion(FunctionContext<String[]> context, Region region,
    GatewaySender sender, long maxRate, int batchSize) {
  final Set<?> entries;
  if (region instanceof PartitionedRegion && sender.isParallel()) {
    entries =
        ((PartitionedRegion) region).getDataStore().getAllLocalBucketRegions()
            .stream()
            .flatMap(br -> ((Set<?>) br.entrySet()).stream()).collect(Collectors.toSet());
  } else {
    entries = region.entrySet();
  }

  final InternalCache cache = (InternalCache) context.getCache();
  int replicatedEntries = 0;
  final long startTime = clock.millis();

  Connection connection = null;
  PoolImpl senderPool = null;
  int batchId = 0;
  try {
    senderPool = ((AbstractGatewaySender) sender).getProxy();
    if (senderPool == null) {
      throw new BatchException70("No connection pool available towards receiver", null, 0,
          batchId);
    }
    connection = senderPool.acquireConnection();
    if (connection == null) {
      throw new BatchException70("No connection available towards receiver", null, 0, batchId);
    }

    Iterator<?> iter = entries.iterator();
    while (iter.hasNext()) {
      int batchIndex = 0;
      List<GatewaySenderEventImpl> batch = new ArrayList<>();
      while (iter.hasNext() && batchIndex < batchSize) {
        GatewaySenderEventImpl event = createGatewaySenderEvent(cache, (InternalRegion) region,
            sender, (Region.Entry) iter.next());
        if (event != null) {
          batch.add(event);
          batchIndex++;
        }
      }
      try {
        sendBatch(sender, connection, senderPool, batchId++,
            replicatedEntries, maxRate, startTime, batch);
        replicatedEntries += batch.size();
      } catch (InterruptedException e) {
        return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
            "Operation canceled after having replicated " + replicatedEntries + " entries");
      } catch (BatchException70 e) {
        return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
            "Error (" + e.getMessage() + ") in operation after having replicated "
                + replicatedEntries + " entries");
      }
    }
  } catch (BatchException70 e) {
    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
        "Error (" + e.getMessage() + ") in operation after having replicated "
            + replicatedEntries + " entries");
  } finally {
    if (connection != null && senderPool != null) {
      ((PooledConnection) connection).setShouldDestroy();
      senderPool.returnConnection(connection);
    }
  }

  return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
      "Entries replicated: " + replicatedEntries);
}
```
```
private void sendBatch(GatewaySender sender, Connection connection, PoolImpl senderPool, int batchId,
    int replicatedEntries, long maxRate, long startTime,
    List batch) throws InterruptedException, BatchException70 {
  GatewaySenderBatchOp.executeOn(connection, senderPool, batch, batchId, false, false);
  GatewaySenderEventRemoteDispatcher.GatewayAck ack =
      (GatewaySenderEventRemoteDispatcher.GatewayAck) GatewaySenderBatchOp.executeOn(connection,
          senderPool);
  if (ack == null) {
    throw new BatchException70("Unknown error sending batch", null, 0, batchId);
  }
  if (ack.getBatchException() != null) {
    throw ack.getBatchException();
  }
  doActionsIfBatchReplicated(startTime, replicatedEntries + batch.size(), maxRate);
}
```

Reply:
Thanks for the code! I have incorporated it with some refactors in a new commit.